x-pack/plugins/alerting/server/task_runner/task_runner.ts

```diff
@@ -386,41 +395,53 @@ export class TaskRunner<
 
         return {
           originalAlerts: alertsCopy,
+          originalRecoveredAlerts: recoveredAlerts,
           updatedRuleTypeState: updatedState || undefined,
           hasReachedAlertLimit: alertFactory.hasReachedAlertLimit(),
         };
       });
 
-    const { activeAlerts, recoveredAlerts } = await this.timer.runWithTimer(
+    const { activeAlerts, recoveredAlerts, currentRecoveredAlerts } = await this.timer.runWithTimer(
       TaskRunnerTimerSpan.ProcessAlerts,
       async () => {
         const {
           newAlerts: processedAlertsNew,
           activeAlerts: processedAlertsActive,
+          currentRecoveredAlerts: processedAlertsRecoveredCurrent,
           recoveredAlerts: processedAlertsRecovered,
         } = processAlerts<State, Context, ActionGroupIds, RecoveryActionGroupId>({
           alerts: this.alerts,
           existingAlerts: originalAlerts,
+          previouslyRecoveredAlerts: originalRecoveredAlerts,
           hasReachedAlertLimit,
           alertLimit: this.maxAlerts,
+          setFlapping: true,
         });
 
+        // the flapping flag will only be determined for active alerts
+        // flapping for recovered alerts will always be set to false
```

Comment:
Glad to see this comment, as I was seeing this happen while testing the code, but it's not true!

Once an alert has gone into flapping: true, it can only go to flapping: false if the flappingHistory is at capacity and is all false.

Thinking through this, I think this implies that we need to keep a new piece of data in the alert state, call it `flapping: boolean`.  It would actually track the flapping state, persistently, and then we can use it in the determination of whether the alert is flapping or not.  

Side effect is that the flapping state will actually be persisted, which will be useful.  I was surprised while running the code to find that it wasn't actually persisted anywhere!

Reply:
Resolved in this commit https://github.com/elastic/kibana/pull/144415/commits/a02d7bb325820ab38295e0b94225c1ec00124e66